### PR TITLE
add support for multiple apps using module federation

### PR DESCRIPTION
**Pull request by hasanayan**
this is a recreated/clean version of https://github.com/getsentry/sentry-webpack-plugin/pull/291

I had to disable a few es-lint rules;
- index.js line 82-83: adding webpack-sources as dependency breaks the tests (some of the tests use older version(s) of node than webpack-sources requires as minimum). On the other hand, if the user is using module federation, their webpack version is at least 5 and webpack-sources come with it. So, adding webpack-sources as a dependency is not a must. For the same reason, I did not put the require statement at the top of the file (it might be missing for users using an older version of webpack)

- index js line 105; it is what it is, the rule says no-underscore-dangle but the object coming from webpack has it 🤷🏼‍♂️

----

TL;DR;

the plugin now injects additional SENTRY_RELEASES variable into global scope as an object that holds versions for all apps keyed by [project]@[org]

-----
When ModuleFederationPlugin is used, assuming the project has a single entry in the webpack config, the build outputs a bundle with two separate entrypoints; usually named as 'main' and 'remoteEntry'. Main is used when the app is running in standalone (host) mode and remoteEntry is used when it is consumed as a remote by another app. In our projects we needed to access the version value for all remotes on our host app so that we can create SentryClient per remote with their own respective versions.

With this PR, the code injected by loader is improved to include a second variable `SENTRY_RELEASES` in addition to `SENTRY_RELEASE`. This code goes to main module and is not loaded when the app is running as remote. So I also added the part to tap into compilation and inject the variable in remoteEntry as well.

If this is accepted, we can maybe open another PR on @sentry/browser so that the default release is read from SENTRY_RELEASES and retire SENTRY_RELEASE

### CODE REVIEW

**Comment by hasanayan** [comment]
@kamilogorek I hope you had a good time in your vacation. I recreated the PR I opened earlier. I had to disable some es-lint rules (on specific lines) so I don't introduce breaking changes (if you think we can break support for earlier versions of node, I can update the PR). could you please run the tests?

File: src/index.js [hunk 1]
```diff
@@ -74,6 +74,58 @@ function attachAfterEmitHook(compiler, callback) {
...
 }
 
+function attachAfterCodeGenerationHook(compiler, options) {
+  if (!compiler.hooks || !compiler.hooks.make) {
+    return;
+  }
+
+  // eslint-disable-next-line global-require
+  const webpackSources = require(// eslint-disable-next-line import/no-extraneous-dependencies
+  'webpack-sources');
```
Comment (kamilogorek) [inline src/index.js:84]:
```suggestion
  let webpackSources;
  try {
    // eslint-disable-next-line global-require, import/no-extraneous-dependencies
    webpackSources = require('webpack-sources');
  } catch(_e) {}
```

**Review by kamilogorek — CHANGES_REQUESTED** [review]
Only one small change. Better safe than sorry.
Thanks for your patience, I got pulled off to other things after coming back.

**Comment by kamilogorek** [comment]
Failing test should be trivial to fix, as its just string based assertion.

**Comment by hasanayan** [comment]
@kamilogorek thanks. I approved your suggestion. I’m on mobile, if there are other changes required I l’ll have a look in a few days. (now is my vacation time and I am on mobile :) )

**Comment by kamilogorek** [comment]
I cannot push to your fork so will have to wait for you. No rush, it won't go anywhere. Enjoy your time off!

File: src/sentry.loader.js [hunk 2]
```diff
@@ -1,8 +1,15 @@
...
-  const { releasePromise } = this.query;
+  const { releasePromise, org, project } = this.query;
   const callback = this.async();
   releasePromise.then(version => {
-    const sentryRelease = `(typeof window !== 'undefined' ? window : typeof global !== 'undefined' ? global : typeof self !== 'undefined' ? self : {}).SENTRY_RELEASE={id:"${version}"};`;
+    let sentryRelease = `const _global = (typeof window !== 'undefined' ? window : typeof global !== 'undefined' ? global : typeof self !== 'undefined' ? self : {}); _global.SENTRY_RELEASE={id:"${version}"};`;
+    if (project) {
+      const key = org ? `${project}@${org}` : project;
+      sentryRelease += `
+      _global.SENTRY_RELEASES=_global.SENTRY_RELEASES||{};
```
Comment (iker-barriocanal) [inline src/sentry.loader.js:9]:
```suggestion
      _global.SENTRY_RELEASES=_global.SENTRY_RELEASES || {};
```

File: example/test.js [hunk 3]
```diff
@@ -8,14 +8,42 @@ console.log();
...
+if (
+  content.toString().indexOf(
+    `_global.SENTRY_RELEASE = {
   id: "foo"
-}`) !== -1
+};`
+  ) !== -1
 ) {
   console.log('Saul Goodman, found SENTRY_RELEASE in bundle');
-  process.exit(0);
```
Comment (iker-barriocanal) [inline example/test.js:17]: Why remove this `process.exit`, and why not add it to the previous check? Should we expect to find all the assignments? If we successfully find one but fail the others, should we exit? If this is the case, should we move the `process.exit(0)` in the last block to run after it?

**Review by iker-barriocanal — COMMENTED** [review]
I rather not merging it since I don't have context, just had a look at the code and can't say whether the bigger picture looks good.

**Comment by iker-barriocanal** [comment]
@hasanayan sorry, I wasn't specific enough. I'm not familiar with this webpack plugin and thus don't want to make the merging decision, which isn't related to your PR at all. Just did a simple review code-wise so I'm definitely not blocking. Updated the comment.

File: example/test.js [hunk 4]
```diff
@@ -8,14 +8,42 @@ console.log();
...
+  id: "foo"
+};`
+  ) !== -1
+) {
+  console.log('Saul Goodman, found SENTRY_RELEASES assignment in bundle');
+  process.exit(0);
+} else {
+  console.error('Boom, did not find SENTRY_RELEASES assignment in bundle');
+  process.exit(1);
+}
```
Comment (kamilogorek) [inline example/test.js:49]:
```suggestion
if (
  content.toString().indexOf(
    `_global.SENTRY_RELEASES = _global.SENTRY_RELEASES || {};
_global.SENTRY_RELEASES["my-project@my-org"] = {
  id: "foo"
};`
  ) !== -1
) {
  console.log('Saul Goodman, found SENTRY_RELEASES assignment in bundle');
} else {
  console.error('Boom, did not find SENTRY_RELEASES assignment in bundle');
  process.exit(1);
}
```

**Review by kamilogorek — APPROVED** [review]

**Comment by hasanayan** [comment]
@kamilogorek sure! on it now

**Comment by dfdeagle47** [comment]
@hasanayan I'm investigating a broken build due to ES6 syntax in our app bundle, so forgive me if my assumption is incorrect, but, won't this line represent a breaking change:

https://github.com/getsentry/sentry-webpack-plugin/commit/e621cea230307ce5bb0f0d33de907313304cfea4#diff-e4d441b957e997f8330cdbc7252040be3195c8af431297cf012aa70c00f66c19R5

The injected Sentry release string contains the `const` keyword:

```js
let sentryRelease = `const _global = (typeof window !== 'undefined' ? window : typeof global !== 'undefined' ? global : typeof self !== 'undefined' ? self : {}); _global.SENTRY_RELEASE={id:"${version}"};`;
```

but the previous code did not contain the `const` keyword in the Sentry release string (if I'm reading the changes right):

```js
const sentryRelease = `(typeof window !== 'undefined' ? window : typeof global !== 'undefined' ? global : typeof self !== 'undefined' ? self : {}).SENTRY_RELEASE={id:"${version}"};`;
```

So if the consumer of this lib doesn't transpile external libs, it could break on browsers which don't support ES6+ syntax.

Happy to provide a PR (and/or create an Issue) if you think the `const` should be changed to `var` for compatibility purposes.

**Comment by kamilogorek** [comment]
You are correct @dfdeagle47, thanks for catching this. Updated in https://github.com/getsentry/sentry-webpack-plugin/pull/338
